viewing_party/party.py

```diff
@@ -1,23 +1,155 @@
-    pass
+    if not title or not genre or not rating:
+        return None
+    return {"title": title, "genre": genre, "rating": rating}
+
+
+def add_to_watched(user_data, movie):
+    user_data["watched"].append(movie)
+    return user_data
+
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data
+
+
+def watch_movie(user_data, title):
+    watchlist = user_data["watchlist"]
+    new_watchlist = []
+
+    for i in range(len(watchlist)):
+        if title == watchlist[i]["title"]:
+            add_to_watched(user_data, watchlist[i])
+        else:
+            new_watchlist.append(watchlist[i])
+    user_data["watchlist"] = new_watchlist
+
+    return user_data
+
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
 
+def get_watched_avg_rating(user_data):
+    if not user_data["watched"]:
+        return 0
+
+    ratings_sum = sum([movie["rating"] for movie in user_data["watched"]])
+    movie_count = len(user_data["watched"])
+
+    return ratings_sum/movie_count
+
+
+def get_most_watched_genre(user_data):
+    if not user_data["watched"]:
+        return None
+
+    genre_count = {}
+
+    for movie in user_data["watched"]:
+        if movie["genre"] in genre_count:
+            genre_count[movie["genre"]] += 1
+        else:
+            genre_count[movie["genre"]] = 1
+
+    return max(genre_count, key=genre_count.get)
+
+
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+
+def get_unique_watched(user_data):
+    friends_watched = []
+    list_of_movies = []
+
+    for i in user_data["friends"]:
+        for j in i["watched"]:
+            friends_watched.append(j["title"])
+
+    for i in user_data["watched"]:
+        if i["title"] not in set(friends_watched):
+            list_of_movies.append(i)
+
+    return list_of_movies
+
+
+def get_friends_unique_watched(user_data):
+    user_titles = []
+    for movie in user_data["watched"]:
+        user_titles.append(movie["title"])
+
+    friends_titles = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_titles.append(movie["title"])
+
+    friends_unique_titles = []
+    for item in set(friends_titles):
+        if item not in set(user_titles):
+            friends_unique_titles.append(item)
+
+    friends_unique_movies = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] in set(friends_unique_titles):
+                friends_unique_movies.append(movie)
+
+    friends_unique_movies_no_duplicates = []
+    for movie in friends_unique_movies:
+        if movie not in friends_unique_movies_no_duplicates:
+            friends_unique_movies_no_duplicates.append(movie)
+
+    return friends_unique_movies_no_duplicates

```

Comment:
I see your approach here, similar to the previous function. You could simplify a lot by using a helper function but even without, you could 
```python
    # This could be extracted into a helper that both this and get_unique_watched could use
    friend_movies = []

    for watched_dict in user_data['friends']:
        for movie in watched_dict['watched']:
            if movie not in friend_movies:
                friend_movies.append(movie)

    # Now you only need to do
    unique_friend_movies = []

    for movie in friend_movies:
        if movie not in user_data['watched']:
            unique_friend_movies.append(movie)

    return unique_friend_movies
```